pgjdbc/src/main/java/org/postgresql/util/PGint4range.java

```diff
@@ -0,0 +1,180 @@
+    setType(TYPE_NAME);
+  }
+
+  /**
+   * Initialize a range with a given bounds.
+   *
+   * @param lowerBound the lower bound, inclusive, {@code null} if not set
+   * @param upperBound the upper bound, exclusive, {@code null} if not set
+   */
+  public PGint4range(@Nullable Integer lowerBound, @Nullable Integer upperBound) {
+    super(lowerBound, upperBound);
+    setType(TYPE_NAME);
+  }
+
+  /**
+   * Required constructor, initializes an empty range.
+   */
+  public PGint4range() {
+    setType(TYPE_NAME);
+  }
+
+  /**
+   * Initialize a range with a given range string representation.
+   *
+   * @param value String represented range (e.g. '[1,3)')
+   * @throws SQLException Is thrown if the string representation has an unknown format
+   * @see #setValue(String)
+   */
+  public PGint4range(@Nullable String value) throws SQLException {
+    super(value);
+    setType(TYPE_NAME);
+  }
+
+  @Override
+  protected String serializeBound(Integer value) {
+    return value.toString();
+  }
+
+  @Override
+  protected Integer parseToken(String token) {
+    return Integer.parseInt(token);
+  }
+
+  @Override
+  public void setByteValue(byte[] value, int offset) throws SQLException {
+    TypeTag tag = TypeTag.valueOf(value[offset]);
+    switch (tag) {
+      case CLOSED: {
+        this.lowerInclusive = true;
+        this.upperInclusive = false;
+
+        this.lowerBound = readLengthTaggedValue(value, offset + 1);
+        this.upperBound = readLengthTaggedValue(value, offset + 9);
+
+        break;
+      }
+      case LOWER_OPEN: {
+        this.lowerInclusive = false;
+        this.upperInclusive = false;
+        this.lowerBound = null;
+
+        this.upperBound = readLengthTaggedValue(value, offset + 1);
+
+        break;
+      }
+      case UPPER_OPEN: {
+        this.lowerInclusive = true;
+        this.upperInclusive = false;
+        this.upperBound = null;
+
+        this.lowerBound = readLengthTaggedValue(value, offset + 1);
+
+        break;
+      }
+      case EMPTY:
+      case BOTH_OPEN: {
+        this.lowerInclusive = false;
+        this.upperInclusive = false;
+        this.lowerBound = null;
+        this.upperBound = null;
+
+        break;
+      }
+
+      default:
+        throw new PSQLException(
+                GT.tr("Conversion to type {0} failed: {1}.", type, value),
+                PSQLState.DATA_TYPE_MISMATCH);
+    }
+  }
+
+  private Integer readLengthTaggedValue(byte[] value, int offset) throws PSQLException {
+    int length = ByteConverter.int4(value, offset);
+    if (length != 4) {
+      throw new PSQLException(
+              GT.tr("Conversion to type {0} failed: {1}.", type, value),
+              PSQLState.DATA_TYPE_MISMATCH);
+    }
+    return ByteConverter.int4(value, offset + 4);
+  }
+
+  @Override
+  public int lengthInBytes() {
+    int length = 1;
+    if (this.isEmpty()) {
+      return length;
+    }
+    if (!this.isLowerInfinite()) {
```

Comment:
While I could go figure out what the nuances around Infinite are, it would be useful for future readers to provide some insight in the comments.

Reply:
I added non-Javadoc comments to the classes explaining the binary representation. In addition I added some comments to the math in the method. Thinking about it maybe the name `#isLowerInfinite` is not ideal and can be improved to better communicate the intention. Would any of the following method have helped you understand the code better:

- `#isLowerBoundSet`
- `#isLowerBoundOpen`
- `#isLowerBoundClosed`
- `#isLowerBounded`
- `#hasLowerBound`

Do you have other proposals?

For reference the [Hibernate Types Range](https://www.javadoc.io/static/com.vladmihalcea/hibernate-types-55/2.14.0/com/vladmihalcea/hibernate/type/range/Range.html) and the [Guava Range](https://javadoc.io/static/com.google.guava/guava/31.0.1-jre/com/google/common/collect/Range.html).